server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentBuilder.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * A fluent builder to create {@code Iterator&lt;ToXContent&gt;} objects
+ */
+public class ChunkedToXContentBuilder implements Iterator<ToXContent> {
+
+    public static ChunkedToXContentBuilder builder(ToXContent.Params params) {
+        return new ChunkedToXContentBuilder(params);
+    }
+
+    private final ToXContent.Params params;
+    private final Stream.Builder<ToXContent> builder = Stream.builder();
```

Comment:
AIUI this will throw an `IllegalStateException` if you attempt to add something after iteration has started. I'd rather that we had an `assert` to make it even clearer that this is not a valid usage pattern.

Reply:
That is deliberate - but the assert is a good idea